### PR TITLE
Added Thickness support for BorderThickness and CornerRadius

### PR DESCRIPTION
**Pull request by Gillibald**
BorderThickness and CornerRadius are now defined by a Thickness. To get a better performance for uniform borders a fallback to old fast rendering is in place.

### CODE REVIEW

File: src/Avalonia.Controls/Border.cs [hunk 1]
```diff
@@ -120,29 +110,205 @@ protected override Size ArrangeOverride(Size finalSize)
...
+                    var innerRect = new Rect(borderThickness.Left, borderThickness.Top,
+                        boundRect.Width - borderThickness.Right, boundRect.Height - borderThickness.Bottom);
+
+                    StreamGeometry backgroundGeometry = null;
+
+                    if (!boundRect.Width.Equals(0) && !boundRect.Height.Equals(0))
+                    {
+                        backgroundGeometry = new StreamGeometry();
+
+                        if (cornerRadius.IsEmpty)
```
Comment (DmitryZhelnin) [inline src/Avalonia.Controls/Border.cs]: Why there are two identical if-else branches?

File: src/Avalonia.Controls/Border.cs [hunk 2]
```diff
@@ -120,29 +110,205 @@ protected override Size ArrangeOverride(Size finalSize)
...
+                    else
+                    {
+                        _backgroundGeometryCache = null;
+                    }
+
+                    if (!boundRect.Width.Equals(0) && !boundRect.Height.Equals(0))
+                    {
+                        var borderGeometry = new StreamGeometry();
+
+                        if (cornerRadius.IsEmpty)
```
Comment (DmitryZhelnin) [inline src/Avalonia.Controls/Border.cs]: And two more identical if-else braches

File: src/Avalonia.Controls/Border.cs [hunk 3]
```diff
@@ -120,29 +110,205 @@ protected override Size ArrangeOverride(Size finalSize)
...
+                {
+                    _useComplexRendering = true;
+
+                    var boundRect = new Rect(finalSize);
+                    var innerRect = new Rect(borderThickness.Left, borderThickness.Top,
+                        boundRect.Width - borderThickness.Right, boundRect.Height - borderThickness.Bottom);
+
+                    StreamGeometry backgroundGeometry = null;
+
+                    if (!boundRect.Width.Equals(0) && !boundRect.Height.Equals(0))
```
Comment (DmitryZhelnin) [inline src/Avalonia.Controls/Border.cs]: Shouldn't it suppose to be `innerRect` inside `if`?

File: src/Avalonia.Controls/Border.cs (hunk 1 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Border.cs]: Good catch. Those checks are not needed anymore.

**Comment by Gillibald** [comment]
Fast rendering needs some improvment dont know how yet. The inner corners dont match if you have a CornerRadius > 0. In my faster rendering path i just take the same radius and it looks smooth.
   ```
if (background != null)
{
        context.FillRectangle(background, rect.Deflate(borders), cornerRadius);
}

if (borderBrush != null && borderThickness > 0)
{
        context.DrawRectangle(new Pen(borderBrush, borderThickness), rect.Deflate(borderThickness), cornerRadius);
}
```
FillRectangle and DrawRectangle seem to produce diffirent rounded corners.

**Comment by Gillibald** [comment]
Microsoft checks for UseLayoutRounding and calculates everything in terms of ui scaling (DPI). Should i do the same and how do i do that? Everything about scaling is private.

**Comment by grokys** [comment]
@Gillibald the layout rounding should have already been applied by the layout pass, so the control should be snapped to pixels. This is done [here](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Layout/Layoutable.cs#L604).

Is it that you want to also snap the border thickness to pixels? To get the layout scaling for a control you can copy this method: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Layout/Layoutable.cs#L698 - it may be something we need to expose publicly, but for the moment just copy it into your codebase and see if it solves your problem, we can look into exposing it once we know it's needed.

**Comment by Gillibald** [comment]
Looks like with the current complex render path all pixels are already snapped to boundaries. I no longer use a Pen in my DrawGeometry call. Don't know why a brush is different but for now it is enough. Only have tested Direct2D maybe other Frameworks behave differently.

File: src/Avalonia.Controls/Presenters/ContentPresenter.cs [hunk 4]
```diff
@@ -372,82 +363,82 @@ private void ContentChanged(AvaloniaPropertyChangedEventArgs e)
...
+
             return finalSize;
         }
 
         private double GetLayoutScale()
         {
             var result = (VisualRoot as ILayoutRoot)?.LayoutScaling ?? 1.0;
 
-            if (result == 0 || double.IsNaN(result) || double.IsInfinity(result))
+            if (result.Equals(0) || double.IsNaN(result) || double.IsInfinity(result))
```
Comment (grokys) [inline src/Avalonia.Controls/Presenters/ContentPresenter.cs]: Why the change here? `result` is a double, so `==` is what you'd usually use to compare.

File: src/Avalonia.Controls/Border.cs [hunk 5]
```diff
@@ -120,29 +110,180 @@ protected override Size ArrangeOverride(Size finalSize)
...
             {
                 child.Measure(availableSize.Deflate(padding));
                 return child.DesiredSize.Inflate(padding);
             }
-            else
+
+            return new Size(padding.Left + padding.Right, padding.Bottom + padding.Top);
+        }
+
+        internal class BorderRenderer
```
Comment (grokys) [inline src/Avalonia.Controls/Border.cs]: I think I'd rather have this class moved out of `Border`. Maybe put it as an internal class under `Utils`?

File: src/Avalonia.Themes.Default/Accents/BaseLight.xaml [hunk 6]
```diff
@@ -20,7 +21,7 @@
     <SolidColorBrush x:Key="ErrorBrush">Red</SolidColorBrush>
     <SolidColorBrush x:Key="ErrorBrushLight">#10ff0000</SolidColorBrush>
 
-    <sys:Double x:Key="ThemeBorderThickness">2</sys:Double>
+    <avalonia:Thickness x:Key="ThemeBorderThickness">2</avalonia:Thickness>
```
Comment (grokys) [inline src/Avalonia.Themes.Default/Accents/BaseLight.xaml]: `Thickness` should be picked up without a namespace. To do this you need to add an `XmlnsDefinition` attribute to the `Avalonia.Visuals` assembly with the plain `Avalonia` namespace. There are already a couple of these definitions: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Properties/AssemblyInfo.cs#L11

File: src/Avalonia.Visuals/CornerRadius.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                default:
+                    {
+                        throw new FormatException("Invalid CornerRadius.");
+                    }
+            }
+        }
+
+        public static bool operator ==(CornerRadius cr1, CornerRadius cr2)
+        {
+            return ((cr1.TopLeft.Equals(cr2.TopLeft) || double.IsNaN(cr1.TopLeft) && double.IsNaN(cr2.TopLeft)))
```
Comment (grokys) [inline src/Avalonia.Visuals/CornerRadius.cs]: What does a `NaN` as a portion of a `CornerRadius` mean exactly?

File: src/Windows/Avalonia.Direct2D1/Media/GeometryImpl.cs [hunk 8]
```diff
@@ -26,8 +26,15 @@ public GeometryImpl(Geometry geometry)
         /// <inheritdoc/>
         public Rect GetRenderBounds(Avalonia.Media.Pen pen)
         {
-            var factory = AvaloniaLocator.Current.GetService<Factory>();
-            return Geometry.GetWidenedBounds((float)pen.Thickness).ToAvalonia();
+            //var factory = AvaloniaLocator.Current.GetService<Factory>();
```
Comment (grokys) [inline src/Windows/Avalonia.Direct2D1/Media/GeometryImpl.cs]: Should probably delete this line rather than commenting it out.

File: src/Windows/Avalonia.Direct2D1/Media/GeometryImpl.cs [hunk 9]
```diff
@@ -26,8 +26,15 @@ public GeometryImpl(Geometry geometry)
...
+            if (pen == null)
+            {
+                return Geometry.GetWidenedBounds(0).ToAvalonia();
+            }
+            else
+            {
+                return Geometry.GetWidenedBounds((float)pen.Thickness).ToAvalonia();
```
Comment (grokys) [inline src/Windows/Avalonia.Direct2D1/Media/GeometryImpl.cs]: Couldn't this block be written more succinctly as: `return Geometry.GetWidenedBounds((float)(pen?.Thickness ?? 0)).ToAvalonia();`?

**Review by grokys — CHANGES_REQUESTED** [review]
Added a few small comments (mainly nits) in the code.

However I think there's something wrong with the complex border drawing:

image

The first border above has a `CornerRadius="16 0 0 0"` and the second `CornerRadius="16"`, however the corner on the top left of the first border doesn't look the same the corners on the second border - it looks "fatter". Any idea why this is?

Also we should probably have some more render tests to test this new functionality.

**Comment by Gillibald** [comment]
Thanks for your feedback will have a look at it soon. Looks like my background geometry needs some tweeks. I probably need to use a percentage of the outer corner radius for the inner. Should be manageable.

File: src/Windows/Avalonia.Direct2D1/Media/GeometryImpl.cs (hunk 9 above)
Comment (Gillibald) [reply src/Windows/Avalonia.Direct2D1/Media/GeometryImpl.cs]: I usually do that wasn't sure about readability and code style. Will have that in mind for the future.

File: src/Avalonia.Controls/Presenters/ContentPresenter.cs (hunk 4 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Presenters/ContentPresenter.cs]: Wasn't sure about rounding errors but I will revert it. Usually it should do the same.

Comment (grokys) [reply src/Avalonia.Controls/Presenters/ContentPresenter.cs]: I'm wasn't aware of any differences between `double.Equals` and `==`, are there differences?!?

File: src/Avalonia.Visuals/CornerRadius.cs (hunk 7 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/CornerRadius.cs]: Good catch. Can't think of a case where we need NaN. I should probably check for NaN and throw an exception. Is NaN allowed for Thickness? Have to check that first.

File: src/Avalonia.Controls/Presenters/ContentPresenter.cs (hunk 4 above)
Comment (grokys) [reply src/Avalonia.Controls/Presenters/ContentPresenter.cs]: As far as I can see, `double.Equals` just calls `==`: https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Double.cs#L171

File: src/Avalonia.Controls/Border.cs (hunk 5 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Border.cs]: Will have a look at present render tests and create new ones to represent my additions. That way we can make sure everything looks as it should be.

**Comment by Gillibald** [comment]
Render tests are still missing. Probably need some help there.

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Apologies for the delay reviewing this @Gillibald - looks good to me now! I added a couple of render tests. Once CI has passed I'll get this merged.

**Comment by Gillibald** [comment]
Thanks for your help. Can we check ContentPresenter before merging. Wasn't sure how to test measurements. ArrangeOverrideImpl has some changes I wasn't sure about.

**Comment by grokys** [comment]
Sure! What in particular were you unsure of?

**Comment by Gillibald** [comment]
In ContentPresenter.cs line 375 - 381

Before my changes just the uniform BordeeThickness took part of all calculations. Now I'm am using all 4 sides.

**Comment by grokys** [comment]
Ok, looks like the best thing there would be to compare with UWP's behavior then. I won't get chance to do that this evening, but I will try to take a look tomorrow!

**Comment by Gillibald** [comment]
Maybe I have to spend some time with that part myself. Padding is handled only with Left and Bottom part. Have to check some ContentPresenter examples and make sure all measurements are fine. Thanks for your help in advance.

**Comment by Gillibald** [comment]
Looks like everything is fine with ContentPresenter. Couldn't find anything unusual. So in my opinion we are ready to merge.

**Comment by grokys** [comment]
@danwalmsley did you give it a test? Are we ok to merge?

**Comment by grokys** [comment]
Thanks @Gillibald ! @danwalmsley has given the 👍 on this in gitter so merged!
